### ISSUE TITLE
IIIF Image API tile source

### ISSUE THREAD

**Issue by lutzhelm** [openlayers/openlayers#9323]
**Is your feature request related to a problem? Please describe.**
For the digital library framework [Kitodo.Presentation](https://github.com/kitodo/kitodo-presentation) that uses OpenLayers to display tiled images from IIP and Zoomify servers as well as static images, I need an additional tile source for [International Image Interoperability Framework](https://iiif.io/) (IIIF) [Image API](https://iiif.io/api/image/) services. Kitodo.Presentation is not only used by several scientific libraries in Germany, it's also the basis for the [DFG Viewer](https://github.com/slub/dfg-viewer), the reference implementation of the German Research Foundation's (DFG) digitization guidelines.

IIIF is a set of API specifications that are primarily implemented and used by GLAM institutions (galleries, libraries, archives, museums and others; see list of active [IIIF community](https://iiif.io/community/#participating-institutions) member institutions) to present digital representations of works such as books, newspapers, scrolls, maps, manuscripts etc. (or other digital images) primarily to scholars and researchers. The IIIF APIs in detail are currently the Image API, Presentation API, Content Search API and Authentication API:
* The [IIIF Presentation API](https://iiif.io/api/presentation/) describes ways to present a digital work to a user, regarding the arrangement of (multiple) images and audio/visual resources, comments, transscriptions, descriptions etc.
* The [IIIF Image API](https://iiif.io/api/image/) describes how to serve/retrieve full or tiled images or specific parts of an image at different scale levels.
* The [IIIF Content Search API](https://iiif.io/api/search/) enables lookup of IIIF resources.
* The [IIIF Authentication API](https://iiif.io/api/auth/) describes how make use of existing authentication services to authenticate users of IIIF services.

In the context of this issue, the API of interest is the IIIF Image API. Support for IIIF would offer an additional way to use custom image tile sources besides IIP and Zoomify. [IIPSrv](https://github.com/ruven/iipsrv), the server side implementation of [IIP](https://iipimage.sourceforge.io/) which is already supported in OpenLayers via the Zoomify tile source does already offer IIIF image support, and there are more IIIF server implementations around.

While implementing an OpenLayers IIIF tile source could be done without changes to OpenLayers itself via a separate project, it would be great to have the support of the large OpenLayers community.

With https://github.com/klokantech/iiifviewer there is already a great IIIF implementation for OpenLayers 4, but it is no longer maintained, and it has several problems (a resolution / zoom level matching bug klokantech/iiifviewer#23, no support for [IIIF level 0 compliant](https://iiif.io/api/image/2.1/compliance/#level-0-compliance) image services, no support for non square tiles, does not use [canonical](https://iiif.io/api/image/2.1/#canonical-uri-syntax) URL syntax).

**Describe the solution you'd like**
I already started [an implementation](https://github.com/lutzhelm/openlayers/blob/iiif/src/ol/source/IIIF.js) which is kind of based on the Zoomify tile source. The linked [branch](https://github.com/lutzhelm/openlayers/tree/iiif) ("iiif") comes with a [working example](https://github.com/lutzhelm/openlayers/blob/iiif/examples/iiif.html). There is still a lot to do (tests, documentation, getting around [rounding issues](https://www.jack-reed.com/2016/10/14/rounding-strategies-used-in-iiif.html)), but it should already support any Image API 1 and 2 services on any compliance level.

I would continue to work on that source and would appreciate any input / feedback.

 @hyperknot mentioned porting the OpenSeaDragon implementation to OpenLayers in #7795. I don't know if they finished this work but it would be an alternative to my suggestion.

**Comment by hyperknot** [comment]
@lutzhelm I extensively reviewed all options, including Klokantech back then, and OpenSeaDragon was by far the best one out there. It's nice code and is actually a full implementation, unlike Klokantech.

Unfortunately the work I did for that company back then isn't open source, so I cannot contribute it to OL, but the code was quite small at the end, something like 200-300 lines maximum. I also skipped some crazy edge cases which is implemented in OSD but realistically no one would ever use it, which helped a lot with the implementation.
